Log throttling

tonyreigns opened the pull request:
- [ ] I have considered the risk of behavior change or other zero migration impact (https://github.com/OpenLiberty/open-liberty/wiki/Behavior-Changes).
- [ ] If this PR fixes an Issue, the description includes "Fixes #FILLMEIN" or "Resolves #FILLMEIN" (verify `release bug` label if applicable: https://github.com/OpenLiberty/open-liberty/wiki/Open-Liberty-Conventions).
- [ ] If this PR resolves an external Known Issue (including APARS), the description includes "Fixes #FILLMEIN" or "Resolves #FILLMEIN".

################################################################################################

Initial drop for #31719

tonyreigns committed:
Initial log throttling drop

Some changes

Moving Log Throttling Logic to bufferMgrImpl

Copying logging.data temporarily to avoid circular dependency

Copying additional logging files temporarily to avoid circular dependency

Move ThrottleState to logging.core

Add garbage collection for throttleStates

Removing logging.core code

Remove migrated coded from logging.core

Move log throttling to BaseTraceService and initial drop

Remove old code

Disable fat simplicity logging methods

Making ID related changes

tonyreigns commented:
#build ([view Open Liberty Personal Build](https://libh-proxy1.fyre.ibm.com/cognitive/pipelineAnalysis.html?pipelineId=8c86e3ea-37cd-4259-989b-b63671c62c7a) - :x: completed with errors/failures.)

_Note: Target locations of links might be accessible only to IBM employees._

tonyreigns commented:
#build ([view Open Liberty Personal Build](https://libh-proxy1.fyre.ibm.com/cognitive/pipelineAnalysis.html?pipelineId=92797f5a-8a42-4ef9-a6cc-906f3021db37) - :x: completed with errors/failures.)

_Note: Target locations of links might be accessible only to IBM employees._

ramkumar-k-9286 commented on dev/com.ibm.ws.logging.osgi/resources/OSGI-INF/l10n/metatype.properties:
```diff
@@ -115,4 +115,12 @@ rollover.interval=Rollover interval for time based log rollover
+throttle.type=Track log entries that match either the messageID or the full message content.
+throttle.type.desc=Track log entries that match either the messageID or the full message content. Specify a string to configure this attribute, the only valid values are message and messageID.
+throttle.type.messageid=Throttle log entries based off the messageID.
+throttle.type.message=Throttle log entries based off the message content.
```
```suggestion
throttle.type.message=Specifies throttling of log entries based on the message content.
```

ramkumar-k-9286 commented on dev/com.ibm.ws.logging.osgi/resources/OSGI-INF/l10n/metatype.properties:
```diff
@@ -115,4 +115,12 @@ rollover.interval=Rollover interval for time based log rollover
+
+throttle.type=Track log entries that match either the messageID or the full message content.
+throttle.type.desc=Track log entries that match either the messageID or the full message content. Specify a string to configure this attribute, the only valid values are message and messageID.
+throttle.type.messageid=Throttle log entries based off the messageID.
```
```suggestion
throttle.type.messageid=Specifies throttling of log entries based on the messageID.
```

ramkumar-k-9286 commented on dev/com.ibm.ws.logging.osgi/resources/OSGI-INF/l10n/metatype.properties:
```diff
@@ -115,4 +115,12 @@ rollover.interval=Rollover interval for time based log rollover
+throttle.max.messages.per.window.desc=Specifies the maximum number of log messages permitted over a 5 minute duration before throttling begins. If this limit is exceeded, any additional messages will be suppressed until the count falls below the threshold. The default is 1000 messages but you can specify a positive integer or set to 0 to disable throttling.
+
+throttle.type=Track log entries that match either the messageID or the full message content.
+throttle.type.desc=Track log entries that match either the messageID or the full message content. Specify a string to configure this attribute, the only valid values are message and messageID.
```
```suggestion
throttle.type.desc=Tracks log entries that match either the messageID or the full message content. Specify a string to configure this attribute. The only valid values are message and messageID.
```

ramkumar-k-9286 commented on dev/com.ibm.ws.logging.osgi/resources/OSGI-INF/l10n/metatype.properties:
```diff
@@ -115,4 +115,12 @@ rollover.interval=Rollover interval for time based log rollover
+throttle.max.messages.per.window=Specifies the maximum number of log messages permitted over a 5 minute duration before throttling begins.
+throttle.max.messages.per.window.desc=Specifies the maximum number of log messages permitted over a 5 minute duration before throttling begins. If this limit is exceeded, any additional messages will be suppressed until the count falls below the threshold. The default is 1000 messages but you can specify a positive integer or set to 0 to disable throttling.
+
+throttle.type=Track log entries that match either the messageID or the full message content.
```
```suggestion
throttle.type=Tracks log entries that match either the messageID or the full message content.
```

ramkumar-k-9286 commented on dev/com.ibm.ws.logging.osgi/resources/OSGI-INF/l10n/metatype.properties:
```diff
@@ -115,4 +115,12 @@ rollover.interval=Rollover interval for time based log rollover
+max.ffdc.age.desc=The maximum wanted age before an FFDC file is deleted. Specify a positive integer followed by a unit of time, which can be days (d), hours (h), or minutes (m). For example, specify 2 days as 2d. You can include multiple values in a single entry. For example, 2d6h is equivalent to 2 days and 6 hours. Everyday at midnight, any FFDC file that reaches the maximum age is deleted. By  ...
+
+throttle.max.messages.per.window=Specifies the maximum number of log messages permitted over a 5 minute duration before throttling begins.
+throttle.max.messages.per.window.desc=Specifies the maximum number of log messages permitted over a 5 minute duration before throttling begins. If this limit is exceeded, any additional messages will be suppressed until the count falls below the threshold. The default is 1000 messages but you can specify a positive integer or set to 0 to disable throttling.
```
```suggestion
throttle.max.messages.per.window.desc=Specifies the maximum number of log messages permitted within a 5-minute period before throttling begins. When the limit is exceeded, additional messages are suppressed until the count is within the threshold. The default value is 1000. You can specify a positive integer or set the value to 0 to disable throttling.
```

ramkumar-k-9286 commented on dev/com.ibm.ws.logging.osgi/resources/OSGI-INF/l10n/metatype.properties:
```diff
@@ -115,4 +115,12 @@ rollover.interval=Rollover interval for time based log rollover
\ No newline at end of file
+max.ffdc.age.desc=The maximum wanted age before an FFDC file is deleted. Specify a positive integer followed by a unit of time, which can be days (d), hours (h), or minutes (m). For example, specify 2 days as 2d. You can include multiple values in a single entry. For example, 2d6h is equivalent to 2 days and 6 hours. Everyday at midnight, any FFDC file that reaches the maximum age is deleted. By  ...
+
+throttle.max.messages.per.window=Specifies the maximum number of log messages permitted over a 5 minute duration before throttling begins.
```
```suggestion
throttle.max.messages.per.window=Specifies the maximum number of log messages that are permitted within 5 minutes before throttling begins.
```

ramkumar-k-9286 commented:
Please apply the corrections to both files.

tonyreigns commented on dev/com.ibm.ws.logging.osgi/resources/OSGI-INF/l10n/metatype.properties:
```diff
@@ -115,4 +115,12 @@ rollover.interval=Rollover interval for time based log rollover
 max.ffdc.age=Maximum age before an FFDC file is deleted
-max.ffdc.age.desc=The maximum wanted age before an FFDC file is deleted. Specify a positive integer followed by a unit of time, which can be days (d), hours (h), or minutes (m). For example, specify 2 days as 2d. You can include multiple values in a single entry. For example, 2d6h is equivalent to 2 days and 6 hours. Everyday at midnight, any FFDC file that reaches the maximum age is deleted. By  ...
\ No newline at end of file
+max.ffdc.age.desc=The maximum wanted age before an FFDC file is deleted. Specify a positive integer followed by a unit of time, which can be days (d), hours (h), or minutes (m). For example, specify 2 days as 2d. You can include multiple values in a single entry. For example, 2d6h is equivalent to 2 days and 6 hours. Everyday at midnight, any FFDC file that reaches the maximum age is deleted. By  ...
```
This is an old attribute that I haven't made any related changes to. Do you still want this changed?

ramkumar-k-9286 replied:
Sorry. You can ignore this.

tonyreigns commented:
@ramkumar-k-9286 All the changes have been made except the ones related to max.ffdc.age.desc, that's an unrelated attribute. Not sure if you want me to still make the change for that attribute.

Thanks!

tonyreigns committed:
Log throttling test cases, betaFence and cleanup

Minor cleanup

Minor cleanup

Add comments

Add beta fence for log throttling

Add additional tests

Resolving code review issues and updating copyright

Fixing Stackoverflow and minor changes

Added error code to tests and added comments for test cases

Remove warning from betaFenceCheck due to stackOverFlow

ID related changes

tonyreigns commented:
#build ([view Open Liberty Personal Build](https://libh-proxy1.fyre.ibm.com/cognitive/pipelineAnalysis.html?pipelineId=4cd354eb-f704-4bdf-9642-4ac110d29083) - :x: completed with errors/failures.)

_Note: Target locations of links might be accessible only to IBM employees._

tonyreigns commented:
#build ([view Open Liberty Personal Build](https://libh-proxy1.fyre.ibm.com/cognitive/pipelineAnalysis.html?pipelineId=4231941e-ff66-4f0d-b6ff-73dcbacc246b) - :x: completed with errors/failures.)

_Note: Target locations of links might be accessible only to IBM employees._

pgunapal commented on dev/com.ibm.ws.logging.osgi/resources/OSGI-INF/metatype/metatype.xml:
```diff
@@ -146,6 +146,22 @@
+        </AD>
+        
+        <AD name="internal" description="internal use only" 
+            id="throttleMapSize" required="false" type="Integer" min="100" default="500">
```
In the UFO, this internal attribute has a bootstrap property and server.env, should we add it in the meta type as well? or should the UFO be updated to only include the server.xml attribute?

pgunapal commented on dev/com.ibm.ws.logging.osgi/resources/OSGI-INF/metatype/metatype.xml:
```diff
@@ -146,6 +146,22 @@
             id="maxFfdcAge" required="false" type="String" ibm:type="duration(m)" default="-1">
         </AD>
 
+        <AD name="%throttle.max.messages.per.window" description="%throttle.max.messages.per.window.desc" 
```
Update copyright year.

pgunapal commented on dev/com.ibm.ws.logging.osgi/resources/OSGI-INF/l10n/metatype.properties:
```diff
@@ -115,4 +115,12 @@ rollover.interval=Rollover interval for time based log rollover
 max.ffdc.age=Maximum age before an FFDC file is deleted
-max.ffdc.age.desc=The maximum wanted age before an FFDC file is deleted. Specify a positive integer followed by a unit of time, which can be days (d), hours (h), or minutes (m). For example, specify 2 days as 2d. You can include multiple values in a single entry. For example, 2d6h is equivalent to 2 days and 6 hours. Everyday at midnight, any FFDC file that reaches the maximum age is deleted. By  ...
\ No newline at end of file
+max.ffdc.age.desc=The maximum wanted age before an FFDC file is deleted. Specify a positive integer followed by a unit of time, which can be days (d), hours (h), or minutes (m). For example, specify 2 days as 2d. You can include multiple values in a single entry. For example, 2d6h is equivalent to 2 days and 6 hours. Everyday at midnight, any FFDC file that reaches the maximum age is deleted. By  ...
```
Update Copyright year.

pgunapal commented on dev/com.ibm.ws.logging/resources/OSGI-INF/l10n/metatype.properties:
```diff
@@ -115,4 +115,12 @@ rollover.interval=Rollover interval for time based log rollover
 max.ffdc.age=Maximum age before an FFDC file is deleted
-max.ffdc.age.desc=The maximum wanted age before an FFDC file is deleted. Specify a positive integer followed by a unit of time, which can be days (d), hours (h), or minutes (m). For example, specify 2 days as 2d. You can include multiple values in a single entry. For example, 2d6h is equivalent to 2 days and 6 hours. Everyday at midnight, any FFDC file that reaches the maximum age is deleted. By  ...
\ No newline at end of file
+max.ffdc.age.desc=The maximum wanted age before an FFDC file is deleted. Specify a positive integer followed by a unit of time, which can be days (d), hours (h), or minutes (m). For example, specify 2 days as 2d. You can include multiple values in a single entry. For example, 2d6h is equivalent to 2 days and 6 hours. Everyday at midnight, any FFDC file that reaches the maximum age is deleted. By  ...
```
Update copyright year.

pgunapal commented on dev/com.ibm.ws.logging/resources/OSGI-INF/metatype/metatype.xml:
```diff
@@ -146,6 +146,22 @@
             id="maxFfdcAge" required="false" type="String" ibm:type="duration(m)" default="-1">
         </AD>
+        
+        <AD name="%throttle.max.messages.per.window" description="%throttle.max.messages.per.window.desc" 
```
Update copyright year.

pgunapal commented on dev/com.ibm.ws.logging/resources/OSGI-INF/metatype/metatype.xml:
```diff
@@ -146,6 +146,22 @@
+        </AD>
+        
+        <AD name="internal" description="internal use only" 
+            id="throttleMapSize" required="false" type="Integer" min="100" default="500">
```
In the UFO, this internal attribute has a bootstrap property and server.env, should we add it in the meta type as well? or should the UFO be updated to only include the server.xml attribute?

pgunapal commented on dev/com.ibm.ws.logging/src/com/ibm/ws/logging/internal/impl/BaseTraceService.java:
```diff
@@ -278,6 +280,20 @@ public static interface TraceWriter extends Closeable {
+    private final Map<String, ThrottleState> throttleStates = new ConcurrentHashMap<>();
+
+    private int throttleMaxMessagesPerWindow = 1000;
+    private final int throttleWindowDurationMS = 5 * 60 * 1000;
```
Good to add a comment to indicate the default duration, for easier readability.

pgunapal commented on dev/com.ibm.ws.logging/src/com/ibm/ws/logging/internal/impl/BaseTraceService.java:
```diff
@@ -1140,33 +1170,150 @@ public void publishLogRecord(LogRecord logRecord) {
-            publishTraceLogRecord(detailLog, logRecord, NULL_ID, formattedMsg, formattedVerboseMsg);
+        if (result == LogResult.THROTTLE_WARNING) {
+            if (routedMessage != null)
+                Tr.warning(tc, "The logs are being throttled due to high volume.");
```
This warning message needs to be translated.

pgunapal commented on dev/com.ibm.ws.logging/src/com/ibm/ws/logging/internal/impl/BaseTraceService.java:
```diff
@@ -1140,33 +1170,150 @@ public void publishLogRecord(LogRecord logRecord) {
+                Tr.warning(tc, "The logs are being throttled due to high volume.");
+
+        } else if (result == LogResult.LOG) {
+
```
formatting issues - new line.

pgunapal commented on dev/com.ibm.ws.logging/src/com/ibm/ws/logging/internal/impl/BaseTraceService.java:
```diff
@@ -1140,33 +1170,150 @@ public void publishLogRecord(LogRecord logRecord) {
+            if (TraceComponent.isAnyTracingEnabled()) {
+                publishTraceLogRecord(detailLog, logRecord, NULL_ID, formattedMsg, formattedVerboseMsg); //One trace
+            }
+
```
formatting issues - new line.

pgunapal commented on dev/com.ibm.ws.logging/src/com/ibm/ws/logging/internal/impl/BaseTraceService.java:
```diff
@@ -1140,33 +1170,150 @@ public void publishLogRecord(LogRecord logRecord) {
+
+        String key;
+
+        if (throttleType.equals("message"))
```
Since throttleType is a String, do we take into consideration for case sensitivity?

pgunapal commented on dev/com.ibm.ws.logging/src/com/ibm/ws/logging/internal/impl/BaseTraceService.java:
```diff
@@ -1140,33 +1170,150 @@ public void publishLogRecord(LogRecord logRecord) {
+            key = logData.getMessageId();
+        }
+
+        if (key == null)
```
Please add comments to explain why happens.

tonyreigns replied:
This should never occur if a log event reaches this point but I added it in case....Already have a check later to see if key is null or not though so removed this.

pgunapal commented on dev/com.ibm.ws.logging/src/com/ibm/ws/logging/internal/impl/BaseTraceService.java:
```diff
@@ -451,6 +468,10 @@ public synchronized void update(LogProviderConfig config) {
             collectorMgrPipelineUtils = CollectorManagerPipelineUtils.getInstance();
         }
 
+        throttleMaxMessagesPerWindow = trConfig.getThrottleMaxMessagesPerWindow();
```
Add beta guard

pgunapal commented on dev/com.ibm.ws.logging/src/com/ibm/ws/logging/internal/impl/BaseTraceService.java:
```diff
@@ -1140,33 +1170,150 @@ public void publishLogRecord(LogRecord logRecord) {
-                publishTraceLogRecord(detailLog, logRecord, NULL_ID, formattedMsg, formattedVerboseMsg);
-                return;
-            }
+            if (throttleMaxMessagesPerWindow > 0) {
```
Add beta guard

pgunapal commented on dev/com.ibm.ws.logging/src/com/ibm/ws/logging/internal/impl/BaseTraceService.java:
```diff
@@ -1140,33 +1170,150 @@ public void publishLogRecord(LogRecord logRecord) {
+
+    }
+
+    public enum LogResult {
```
Good to add comments to explain each ENUM value.

pgunapal commented on dev/com.ibm.ws.logging/src/com/ibm/ws/logging/internal/impl/BaseTraceService.java:
```diff
@@ -1140,33 +1170,150 @@ public void publishLogRecord(LogRecord logRecord) {
+    public enum LogResult {
+        LOG,
+        THROTTLE,
+        THROTTLE_WARNING
```
From my understanding, this ENUM only checks if it should LOG or a THROTTLE the message, as the result. I find adding the check if we should log the throttling warning message within the ENUM logResult, is kind of confusing. Can we add the check if the log throttling warning message is already printed, outside of this ENUM?

pgunapal commented on dev/com.ibm.ws.logging/src/com/ibm/ws/logging/internal/impl/BaseTraceService.java:
```diff
@@ -1140,33 +1170,150 @@ public void publishLogRecord(LogRecord logRecord) {
+
+                if (shouldSupress) {
+                    if (canLogThrottledMessage()) {
+                        return LogResult.THROTTLE_WARNING;
```
Maybe add an atomic boolean to check if the throttle warning is already printed, instead of ENUM to avoid confusion?

pgunapal commented on dev/com.ibm.ws.logging.osgi/resources/OSGI-INF/metatype/metatype.xml:
```diff
@@ -146,6 +146,22 @@
+            <Option label="%throttle.type.message" value="message"/>
+        </AD>
+        
+        <AD name="internal" description="internal use only" 
```
The UFO has server ENV and bootstrap properties for this internal attribute, should it be added here as well? or should the UFO be updated to remove the env and prop, to match the other existing internal attributes for logging?

tonyreigns replied:
No, it's not needed. The UFO will be updated.

pgunapal commented on dev/com.ibm.ws.logging/resources/OSGI-INF/metatype/metatype.xml:
```diff
@@ -146,6 +146,22 @@
+            <Option label="%throttle.type.message" value="message"/>
+        </AD>
+        
+        <AD name="internal" description="internal use only" 
```
The UFO has server ENV and bootstrap properties for this internal attribute, should it be added here as well? or should the UFO be updated to remove the env and prop, to match the other existing internal attributes for logging?

tonyreigns replied:
No, it's not needed. The UFO will be updated.

pgunapal commented on dev/com.ibm.ws.logging/src/com/ibm/ws/logging/internal/impl/LogProviderConfigImpl.java:
```diff
@@ -4,7 +4,7 @@
  * which accompanies this distribution, and is available at
  * http://www.eclipse.org/legal/epl-2.0/
- * 
+ *
```
Update copyright year

pgunapal commented on dev/com.ibm.ws.logging/src/com/ibm/ws/logging/internal/impl/LogProviderConfigImpl.java:
```diff
@@ -578,7 +608,10 @@ private static enum InitConfgAttribute {
+        NEW_LOGS_ON_START("newLogsOnStart", FileLogHolder.NEW_LOGS_ON_START_PROPERTY),
+        THROTTLE_MAX_MESSAGES_PER_WINDOW("throttleMaxMessagesPerWindow", "com.ibm.ws.logging.throttle.max.messages.per.window"),
+        THROTTLE_TYPE("throttleType", "com.ibm.ws.logging.throttle.type"),
+        THROTTLE_MAP_SIZE("throttleMapSize", "com.ibm.ws.logging.throttle.map.size");
```
The bootstrap property is not set in the meta type. Refer to the meta type.xml for concerns regarding the internal throttle map size attribute.

pgunapal commented on dev/com.ibm.ws.logging/src/com/ibm/ws/logging/internal/impl/LoggingConstants.java:
```diff
@@ -4,7 +4,7 @@
  * which accompanies this distribution, and is available at
  * http://www.eclipse.org/legal/epl-2.0/
- * 
+ *
```
Update copyright year

pgunapal commented on dev/com.ibm.ws.logging/src/com/ibm/ws/logging/internal/impl/MessageLogHandler.java:
```diff
@@ -4,7 +4,7 @@
  * which accompanies this distribution, and is available at
  * http://www.eclipse.org/legal/epl-2.0/
- *
+ * 
```
This file has no relevant changes/updates, please remove it from PR to avoid unnecessary commit history changes to the file.

pgunapal commented on dev/com.ibm.ws.logging/src/com/ibm/ws/logging/internal/impl/ThrottleState.java:
```diff
@@ -0,0 +1,98 @@
+
+import java.util.function.Supplier;
+
+class ThrottleState {
```
Add the `public` access modifier here to be consistent with the other classes.

pgunapal replied:
Add a comment on what this ThrottleState class is doing, and explain the moving window concept.

pgunapal commented on dev/com.ibm.ws.logging_2_fat/fat/src/com/ibm/ws/logging/fat/LogThrottleTest.java:
```diff
@@ -0,0 +1,159 @@
+/*******************************************************************************
+ * Copyright (c) 2022, 2023 IBM Corporation and others.
```
New file, should only be 2025.

pgunapal commented on dev/com.ibm.ws.logging_2_fat/fat/src/com/ibm/ws/logging/fat/LogThrottleTest.java:
```diff
@@ -0,0 +1,159 @@
+    @After
+    public void cleanUp() throws Exception {
+        if (server != null && server.isStarted()) {
+            server.stopServer("com.ibm.ws.logging.fat.ffdc.servlet.FFDCServlet.doGet", "ArithmeticException",
```
I don't think FAT app tests any FFDC?

pgunapal commented on dev/com.ibm.ws.logging_2_fat/fat/src/com/ibm/ws/logging/fat/LogThrottleTest.java:
```diff
@@ -0,0 +1,159 @@
+    public void cleanUp() throws Exception {
+        if (server != null && server.isStarted()) {
+            server.stopServer("com.ibm.ws.logging.fat.ffdc.servlet.FFDCServlet.doGet", "ArithmeticException",
+                              "CWWKG0081E", "CWWKG0083W", "TRAS3015W", "TRAS3013W");
```
Are these ignored messageIDs necessary for this FAT?

pgunapal commented on dev/com.ibm.ws.logging_2_fat/fat/src/com/ibm/ws/logging/fat/LogThrottleTest.java:
```diff
@@ -0,0 +1,159 @@
+
+        RemoteFile messagesLogFile = server.getDefaultLogFile();
+        String line = server.waitForStringInLog("The logs are being throttled due to high volume.", 5000, messagesLogFile);
+        assertTrue("The FFDC file was not deleted.", line.contains("The logs are being throttled due to high volume."));
```
The assert message doesn't relate to this test.

pgunapal commented on dev/com.ibm.ws.logging_2_fat/fat/src/com/ibm/ws/logging/fat/LogThrottleTest.java:
```diff
@@ -0,0 +1,159 @@
+    }
+
+    @Test
+    public void testLogThrottlingWarningTriggered() throws Exception {
```
Should we have test case to test if the warning is logged only once?

pgunapal replied:
Add comments to each test case to explain what each test is doing.

pgunapal commented on dev/com.ibm.ws.logging_2_fat/publish/files/server-highMaxMessages.xml:
```diff
@@ -0,0 +1,23 @@
+<!--
+    Copyright (c) 2021 IBM Corporation and others.
```
Update copyright year

pgunapal commented on dev/com.ibm.ws.logging_2_fat/fat/src/com/ibm/ws/logging/fat/LogThrottleTest.java:
```diff
@@ -0,0 +1,159 @@
+        assertTrue("Log Throttling was not disabled",
+                   line.contains("A validation failure occurred while processing the [throttleMaxMessagesPerWindow] property, value = [-1]. Default value in use: [1000]."));
+    }
+
```
Should you add a test for throttleMapSize?

tonyreigns replied:
If we're able to add debug, I can try adding a test.

pgunapal commented on dev/com.ibm.ws.logging_2_fat/fat/src/com/ibm/ws/logging/fat/LogThrottleTest.java:
```diff
@@ -0,0 +1,159 @@
+        assertTrue("Log Throttling was not disabled",
+                   line.contains("A validation failure occurred while processing the [throttleMaxMessagesPerWindow] property, value = [-1]. Default value in use: [1000]."));
+    }
+
```
Should add a test for the equivalent bootstrap property and ENV values of the supported attributes.

tonyreigns replied:
This will be done after the initial drop is merged.

pgunapal commented on dev/com.ibm.ws.logging_2_fat/fat/src/com/ibm/ws/logging/fat/LogThrottleTest.java:
```diff
@@ -0,0 +1,159 @@
+                   line.contains("A validation failure occurred while processing the [throttleMaxMessagesPerWindow] property, value = [-1]. Default value in use: [1000]."));
+    }
+
+    public void spamServerConfigurationUpdates(int numberOfUpdates) throws Exception {
```
Should add a test for throttleType case sensitive/insensitive and invalid values.

tonyreigns replied:
Already done to test invalid values. I can add one to test case sensitive input.

pgunapal commented on dev/com.ibm.ws.logging_2_fat/publish/files/server-invalidThrottlingConfig.xml:
```diff
@@ -0,0 +1,23 @@
+<!--
+    Copyright (c) 2021 IBM Corporation and others.
```
Update copyright year.

pgunapal commented on dev/com.ibm.ws.logging_2_fat/fat/src/com/ibm/ws/logging/fat/LogThrottleTest.java:
```diff
@@ -0,0 +1,159 @@
+        assertEquals("Configuration updated message wasn't printed the correct number of times.", lines2.size(), 8); //This message shouldn't be getting throttled due to message variation
+    }
+
+    @Test
```
Is this feature always enabled? We should add a test without the new attributes set and ensure its working with the default values.

tonyreigns replied:
It's enabled by default, to disable you set throttleMaxMessagesPerWindow to 0. The default is 1000 messages so it might be a bit excessive to test but I'll add an additional test later to test this.

pgunapal commented on dev/com.ibm.ws.logging_2_fat/publish/files/server-throttlingFullMessage.xml:
```diff
@@ -0,0 +1,23 @@
+<!--
+    Copyright (c) 2021 IBM Corporation and others.
```
Update copyright

pgunapal commented on dev/com.ibm.ws.logging_2_fat/publish/servers/com.ibm.ws.logging.logThrottle/bootstrap.properties:
```diff
@@ -0,0 +1,13 @@
+###############################################################################
+# Copyright (c) 2021 IBM Corporation and others.
```
Update copyright

pgunapal commented on dev/com.ibm.ws.logging_2_fat/publish/servers/com.ibm.ws.logging.logThrottle/server.xml:
```diff
@@ -0,0 +1,23 @@
+<!--
+    Copyright (c) 2021 IBM Corporation and others.
```
Update copyright year.

pgunapal commented on dev/fattest.simplicity/src/com/ibm/websphere/simplicity/config/Logging.java:
```diff
@@ -4,7 +4,7 @@
  * are made available under the terms of the Eclipse Public License 2.0
```
No meaningful change, please remove file from PR.

pgunapal reviewed (changes requested):
Added some comments, it would be good if you add some trace to where ever you feel necessary that wont interfere with the throttling, maybe when you get the configuration. Also, did you test this with the hideMessage attribute and see if everything works fine.

tonyreigns commented:
#build ([view Open Liberty Personal Build](https://libh-proxy1.fyre.ibm.com/cognitive/pipelineAnalysis.html?pipelineId=f7f45cc8-b030-4dcf-b358-dba70a291c06) - :x: completed with errors/failures.)

_Note: Target locations of links might be accessible only to IBM employees._

tonyreigns commented:
#build ([view Open Liberty Personal Build](https://libh-proxy1.fyre.ibm.com/cognitive/pipelineAnalysis.html?pipelineId=0bcf174c-dfef-41a9-b335-5bb5be8b5bee) - :x: completed with errors/failures.)

_Note: Target locations of links might be accessible only to IBM employees._

tonyreigns commented:
#build ([view Open Liberty Personal Build](https://libh-proxy1.fyre.ibm.com/cognitive/pipelineAnalysis.html?pipelineId=621fcdda-c2c0-4f95-bc0e-d77113b66043) - :warning: completed with warnings.)

_Note: Target locations of links might be accessible only to IBM employees._

ramkumar-k-9286 commented on dev/com.ibm.ws.logging/resources/com/ibm/ws/logging/internal/resources/LoggingMessages.nlsprops:
```diff
@@ -170,6 +170,10 @@ JSON_ACCESS_LOG_NO_HEADER_NAME_SPECIFIED=TRAS3012W: Header name is missing for o
 
+LOG_THROTTLING_ACTIVE_WARNING=TRAS3016W: The logs are being throttled due to high volume.
+LOG_THROTTLING_ACTIVE_WARNING.explanation=Some logs are currently being throttled due to high volume.
+LOG_THROTTLING_ACTIVE_WARNING.useraction=Resolve any issues causing excessive logging or increase logging threshold by configuring the throttleMaxMessagesPerWindow logging attribute. 
```
@copilot 
```suggestion
LOG_THROTTLING_ACTIVE_WARNING.useraction=Resolve any issues that cause excessive logging or increase the logging threshold by configuring the throttleMaxMessagesPerWindow logging attribute. 
```

ramkumar-k-9286 commented on dev/com.ibm.ws.logging/resources/com/ibm/ws/logging/internal/resources/LoggingMessages.nlsprops:
```diff
@@ -170,6 +170,10 @@ JSON_ACCESS_LOG_NO_HEADER_NAME_SPECIFIED=TRAS3012W: Header name is missing for o
 JSON_ACCESS_LOG_NO_HEADER_NAME_SPECIFIED.useraction=Provide a header name for the field in the accessLogging logFormat attribute.
 
+LOG_THROTTLING_ACTIVE_WARNING=TRAS3016W: The logs are being throttled due to high volume.
+LOG_THROTTLING_ACTIVE_WARNING.explanation=Some logs are currently being throttled due to high volume.
```
```suggestion
LOG_THROTTLING_ACTIVE_WARNING.explanation=The logs are currently being throttled due to high volume.
```

pgunapal commented on dev/com.ibm.ws.logging.osgi/resources/OSGI-INF/l10n/metatype.properties:
```diff
@@ -115,4 +115,12 @@ rollover.interval=Rollover interval for time based log rollover
\ No newline at end of file
+max.ffdc.age.desc=The maximum wanted age before an FFDC file is deleted. Specify a positive integer followed by a unit of time, which can be days (d), hours (h), or minutes (m). For example, specify 2 days as 2d. You can include multiple values in a single entry. For example, 2d6h is equivalent to 2 days and 6 hours. Everyday at midnight, any FFDC file that reaches the maximum age is deleted. By  ...
+
+throttle.max.messages.per.window=Specifies the maximum number of log messages that are permitted within 5 minutes before throttling begins
```
Should this sentence end with a period?

tonyreigns replied:
No, you can't have a trailing period for the property.

pgunapal commented on dev/com.ibm.ws.logging.osgi/resources/OSGI-INF/l10n/metatype.properties:
```diff
@@ -115,4 +115,12 @@ rollover.interval=Rollover interval for time based log rollover
+throttle.max.messages.per.window=Specifies the maximum number of log messages that are permitted within 5 minutes before throttling begins
+throttle.max.messages.per.window.desc=Specifies the maximum number of log messages permitted within a 5-minute period before throttling begins. When the limit is exceeded, additional messages are suppressed until the count is within the threshold. The default value is 1000. You can specify a positive integer or set the value to 0 to disable throttling.
+
+throttle.type=Tracks log entries that match either the messageID or the full message content
```
Should this sentence end with a period?

pgunapal commented on dev/com.ibm.ws.logging/resources/OSGI-INF/l10n/metatype.properties:
```diff
@@ -115,4 +115,12 @@ rollover.interval=Rollover interval for time based log rollover
\ No newline at end of file
+max.ffdc.age.desc=The maximum wanted age before an FFDC file is deleted. Specify a positive integer followed by a unit of time, which can be days (d), hours (h), or minutes (m). For example, specify 2 days as 2d. You can include multiple values in a single entry. For example, 2d6h is equivalent to 2 days and 6 hours. Everyday at midnight, any FFDC file that reaches the maximum age is deleted. By  ...
+
+throttle.max.messages.per.window=Specifies the maximum number of log messages that are permitted within 5 minutes before throttling begins
```
Does this sentence need a ending period?

pgunapal commented on dev/com.ibm.ws.logging/resources/OSGI-INF/l10n/metatype.properties:
```diff
@@ -115,4 +115,12 @@ rollover.interval=Rollover interval for time based log rollover
+throttle.max.messages.per.window=Specifies the maximum number of log messages that are permitted within 5 minutes before throttling begins
+throttle.max.messages.per.window.desc=Specifies the maximum number of log messages permitted within a 5-minute period before throttling begins. When the limit is exceeded, additional messages are suppressed until the count is within the threshold. The default value is 1000. You can specify a positive integer or set the value to 0 to disable throttling.
+
+throttle.type=Tracks log entries that match either the messageID or the full message content
```
Does this sentence need a ending period?

pgunapal reviewed (changes requested):
minor comments.

pgunapal reviewed (approved):
LGTM - issue (https://github.com/OpenLiberty/open-liberty/issues/33272) opened for follow-on items discussed in this PR.
